lib/rbac.ex

```diff
@@ -37,4 +37,117 @@ defmodule RBAC do
   def transform_role_list_to_string(roles) do
     [Map.delete(roles, :__meta__)] |> transform_role_list_to_string()
   end
+
+  @doc """
+  `get_approles/2` fetches the roles for the app
+  """
+  def get_approles(auth_url, client_id) do
+    url = "#{auth_url}/approles/#{client_id}"
+    HTTPoison.start()
+
+    HTTPoison.get(url)
+    |> parse_body_response()
+  end
+
+  @doc """
+  `parse_body_response/1` parses the response
+  so your app can use the resulting JSON (list of roles).
+  """
+  @spec parse_body_response({atom, String.t()}) :: String.t()
+  def parse_body_response({:error, err}), do: {:error, err}
+
+  def parse_body_response({:ok, response}) do
+    body = Map.get(response, :body)
+    # IO.inspect(body)
+    # make keys of map atoms for easier access in templates
+    if body == nil do
+      {:error, :no_body}
+    else
+      {:ok, str_key_map} = Jason.decode(body)
+
+      atom_key_map =
+        Enum.map(str_key_map, fn role ->
+          for {key, val} <- role, into: %{}, do: {String.to_atom(key), val}
+        end)
+
+      {:ok, atom_key_map}
+    end
+
+    # https://stackoverflow.com/questions/31990134
+  end
+
+  @doc """
+  `init_roles/2 fetches the list of roles for an app 
+  from the auth app (auth_url) based on the client_id
+  and caches the list for fast access. 
+  ETS is an in-memory cache you get for *Free* in Elixir/Erlang.
+  See: https://elixir-lang.org/getting-started/mix-otp/ets.html
+  and: https://elixirschool.com/en/lessons/specifics/ets
+  """
+  def init_roles_cache(auth_url, client_id) do
+    {:ok, roles} = RBAC.get_approles(auth_url, client_id)
+    :ets.new(:roles_cache, [:set, :protected, :named_table])
+    # insert full list:
+    :ets.insert(:roles_cache, {"roles", roles})
+    # insert individual roles for fast lookup:
+    Enum.each(roles, fn role ->
+      :ets.insert(:roles_cache, {role.name, role})
+      :ets.insert(:roles_cache, {role.id, role})
+    end)
+  end
+
+  @doc """
+  `get_role_from_cache/1 retrieves a role from ets cache
+  """
+  def get_role_from_cache(term) do
+    case :ets.lookup(:roles_cache, term) do
+      # not found:
+      [] -> %{id: 0}
```

Comment:
I think returning an explicit error here instead of a zero value will be clearer

Reply:
Yeah, this is a **_silent_ fail** when the role is not found. 🤭 
It's probably preferable to give the developer _some_ indication that they typoed the role name. 👍

Reply:
I've added a `Logger.error` with the stack trace so devs will be informed of a typo in role. 📣
e.g: 
```
09:44:39.356 [error] rbac.ex:112 Role not found in ets cache: admni
```
The way the cache miss does not fail _silently_ so the DevX doesn't suck. 
Very happy for this to be refactored to return an `:error` when the role is not found, 
and for that to be handled by the calling function. 👍 
But simply updating the code to `:error` (_and breaking the tests_) 
without handling the `:error` is counter-productive. 💭

Reply:
Please open an issue to refactor this if you feel that way inclined. 
Hopefully the `has_role?/2` function is already useful in completing your mission. 👍